bots: Fix edit bot avatar settings.

Ddharmani3 opened the pull request:
Fixed gravatar not visible in edit bot form by making bot_avatar_url equal to bot.avatar_url || people.medium_avatar_url_for_person(bot) which solves the case when avatar_url is returned null.

Added new feature for completely removing user uploaded avatar by adding a delete button which shows on hovering on current avatar.

I added a delete button that appears when hovering over the current avatar. 
Then, I added an event listener for the delete button in the build_bot_edit_widget function.
When clicked, it changes the current avatar image to the bot's gravatar using people.medium_gravatar_url_for_email(), and sets the value of a temporary variable bot_details.avatar_source to "G". 
This temporary avatar_source is only passed to the backend when the user confirms the change by clicking the dialog_submit_button in the edit bot form.

Updated format_user_row function in lib/users.py to also return avatar_source field in realm_user. Updated zulip.yaml and 4 other tests to incorporate the new field avatar_source in realm_user.

Follow-up:  [realm_bots: Only send avatar_url when client cannot compute it. #24698](https://github.com/zulip/zulip/issues/24698)

<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->
#24427

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

![image](https://user-images.githubusercontent.com/64723994/225087181-14216e6c-f21d-444f-9548-d63a95671904.png)

![image](https://user-images.githubusercontent.com/64723994/225350026-6eaceb73-d7a8-4114-bbb5-c6d580cc885e.png)

![image](https://user-images.githubusercontent.com/64723994/225087209-f71f24eb-7ee3-4a7e-907b-4190caf8a5e0.png)


https://user-images.githubusercontent.com/64723994/225087419-0bdf357d-1f1b-4683-a0d8-0244ab0b793e.mp4
<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Ddharmani3 commented:
@sahil839, I have addressed both issues stated in #24427. As per your suggestions [here](https://github.com/zulip/zulip/issues/24427#issuecomment-1467308694), I tackled the issue of avatar_url being null for realm_user by calculating the gravatar for the user using `people.medium_avatar_url_for_person()` when bot.avatar_url is `null`, using the or operator: `bot_avatar_url = bot.avatar_url || people.medium_avatar_url_for_person(bot)`.

To delete the current avatar, I used the approach shared in the PR message above. Please let me know if you like this approach or if you have any other suggestions.

Currently, I am updating the avatar_source in the frontend using `bot.avatar_source = result.avatar_source`, where result is the response received on success of the patch request on the endpoint `bots/{bot_id}`. This was necessary as the avatar_source was not updating in the frontend until the page was reloaded. I have asked for a solution to this issue on [CZO](https://chat.zulip.org/#narrow/stream/21-provision-help/topic/patch_bot_backend/near/1523594). If you have any suggestions, please feel free to share them.

sahil839 commented:
@Ddharmani3 This PR is currently hard to review. You can update the PR to have two different commits - one for just showing the avatar in edit modal and second to allow deleting the profile picture.

sahil839 commented:
> @sahil839 after receiving suggestions in CZO and reading [Real-time push and events](https://zulip.readthedocs.io/en/latest/subsystems/events-system.html). I tried these changes:
> 
> ```
> @transaction.atomic(durable=True)
> def do_change_avatar_fields(
>     user_profile: UserProfile,
>     avatar_source: str,
>     skip_notify: bool = False,
>     *,
>     acting_user: Optional[UserProfile],
> ) -> None:
>     user_profile.avatar_source = avatar_source
>     user_profile.avatar_version += 1
>     user_profile.save(update_fields=["avatar_source", "avatar_version"])
>     event_time = timezone_now()
>     RealmAuditLog.objects.create(
>         realm=user_profile.realm,
>         modified_user=user_profile,
>         event_type=RealmAuditLog.USER_AVATAR_SOURCE_CHANGED,
>         extra_data=str({"avatar_source": avatar_source}),
>         event_time=event_time,
>         acting_user=acting_user,
>     )
>     if user_profile.is_bot:
>         bot_event = dict(
>             type="realm_bot",
>             op="update",
>             bot=dict(
>                 user_id=user_profile.id,
>                 avatar_source=avatar_source,
>             ),
>         )
>         transaction.on_commit(
>             lambda: send_event(
>                 user_profile.realm,
>                 bot_event,
>                 bot_owner_user_ids(user_profile),
>             )
>         )
>     event = dict(
>         type="realm_user",
>         op="update",
>         person=dict(
>             user_id=user_profile.id,
>             avatar_source=avatar_source,
>         ),
>     )
>     transaction.on_commit(
>         lambda: send_event(
>             user_profile.realm,
>             event,
>             active_user_ids(user_profile.realm_id),
>         )
>     )
> 
>     if not skip_notify:
>         notify_avatar_url_change(user_profile)
> ```
> 
> but still, I am unable to update the avatar_source in frontend from the backend in real time. can you please help me with this?

It is not clear what you have changed from the existing code here. It is always better to share changes in a diff format.

Ddharmani3 committed:
bots: Fix gravatar not visible in edit bot modal.

Fixed gravatar not visible in edit bot form by making bot_avatar_url
equal to bot.avatar_url || people.medium_avatar_url_for_person(bot)
which solves the case when avatar_url is returned null.

Fixes #24427

Ddharmani3 commented:
> It is not clear what you have changed from the existing code here. It is always better to share changes in a diff format.
```diff
-@transaction.atomic(savepoint=False)
+@transaction.atomic(durable=True)
def do_change_avatar_fields(
    user_profile: UserProfile,
    avatar_source: str,
    skip_notify: bool = False,
    *,
    acting_user: Optional[UserProfile],
) -> None:
    user_profile.avatar_source = avatar_source
    user_profile.avatar_version += 1
    user_profile.save(update_fields=["avatar_source", "avatar_version"])
    event_time = timezone_now()
    RealmAuditLog.objects.create(
        realm=user_profile.realm,
        modified_user=user_profile,
        event_type=RealmAuditLog.USER_AVATAR_SOURCE_CHANGED,
        extra_data=str({"avatar_source": avatar_source}),
        event_time=event_time,
        acting_user=acting_user,
    )
+    if user_profile.is_bot:
+      bot_event = dict(
+            type="realm_bot",
+           op="update",
+            bot=dict(
+                user_id=user_profile.id,
+                avatar_source=avatar_source,
+            ),
+        )
+        transaction.on_commit(
+            lambda: send_event(
+                user_profile.realm,
+                bot_event,
+                bot_owner_user_ids(user_profile),
+          )
+        )
+    event = dict(
+        type="realm_user",
+      op="update",
+        person=dict(
+            user_id=user_profile.id,
+          avatar_source=avatar_source,
+        ),
+   )
+    transaction.on_commit(
+        lambda: send_event(
+          user_profile.realm,
+           event,
+            active_user_ids(user_profile.realm_id),
+      )
+    )

    if not skip_notify:
        notify_avatar_url_change(user_profile)
```

Ddharmani3 commented:
@sahil839, I have split the PR into two commits. The first commit addresses the issue of showing the avatar in the edit modal when the avatar_url is null. This commit has only one line and should be merged easily.

The second commit allows removing the current avatar if it is not a Gravatar. This commit does the work as expected, but there is one line in this commit where I am updating the avatar_source in the frontend using bot.avatar_source = result.avatar_source. Here, the result is the response received on the success of the patch request on the endpoint bots/{bot_id}. The reason for this update is that the frontend was not auto-updating until a page reload, while the backend was updating properly. However, I am not sure if updating the realm_user's avatar_source in the frontend could cause any bugs in the future.

I have also attempted to make changes in the backend to update the avatar_source in real-time. I have added the code diff above to show the changes I made. However, these changes did not work as expected. Can you review the code changes and suggest any modifications? If I am completely off-track, I can create an issue to solve the real-time updating problem, which can be addressed by someone who is more knowledgeable about Django backend after this PR is merged.

sahil839 commented on zerver/views/users.py:
```diff
@@ -412,7 +413,7 @@ def patch_bot_backend(
         do_update_bot_config_data(bot, config_data)
 
     if len(request.FILES) == 0:
-        pass
```
I don't think this is a correct way to do this. We should instead call this if `avatar_source` value is sent and is equal to `AVATAR_FROM_GRAVATAR` with previous value being `AVATAR_FROM_USER`.

Ddharmani3 replied:
addressed this in the updated PR.

sahil839 commented on web/src/avatar.js:
```diff
@@ -34,11 +35,32 @@ export function build_bot_create_widget() {
+        $("#edit-bot-avatar-upload-widget .image-hover-background").hide();
+        $("#edit-bot-avatar-upload-widget .image-delete-button").hide();
+
+        bot_details.avatar_source = "G";
```
Instead of doing, we can probably have a hidden input with stores value of avatar source. Mutating the object passed to function is not the best solution I think.

Ddharmani3 replied:
fixed this in the updated PR.

sahil839 commented on zerver/views/users.py:
```diff
@@ -425,6 +426,7 @@ def patch_bot_backend(
 
     json_result = dict(
         full_name=bot.full_name,
+        avatar_source=avatar_source,
```
I don't think it is necessary to pass `avatar_source` here in the result dict. We don't do so for human users as well.

Ddharmani3 replied:
I was passing this in the result just to update avatar_source in the frontend. If avatar_source updates in the frontend in real-time, this won't be needed anymore.

sahil839 replied:
Ok, Maybe you can try the solution mentioned [here](https://github.com/zulip/zulip/pull/24697#issuecomment-1482536795).

sahil839 commented on web/src/avatar.js:
```diff
@@ -34,11 +35,32 @@ export function build_bot_create_widget() {
+        e.stopPropagation();
+        $("#edit-bot-avatar-upload-widget .upload-spinner-background").css({visibility: "visible"});
+        $("#edit-bot-avatar-upload-widget .image-hover-background").hide();
+        $("#edit-bot-avatar-upload-widget .image-delete-button").hide();
```
This duplication can be avoided by refactoring `display_avatar_delete_complete` to accept a container element as parameter and we can pass `$("#user-avatar-upload-widget")` and `$("#edit-bot-avatar-upload-widget")` to it. This should be done in a separate prep commit.

Ddharmani3 replied:
I will create a follow-up pull request to display a loading spinner in the bot edit widget when the current avatar is deleted. Just refactoring the `display_avatar_delete_complete` function won't work, because for the profile picture, it was called on the success of the '/json/users/me/avatar' endpoint's call. However, in the bot avatar widget, when the delete button is clicked, only the source of the image element of `#current_bot_avatar_image` is changed to the gravatar URL. I will tackle this by making it asynchronous and calling `display_avatar_delete_complete` once the image source is updated.

or Should I do this in a separate commit within this pull request?

sahil839 replied:
I am not sure whether we would need a spinner here since the request to delete the avatar will be sent only after clicking "Save" button as you mentioned, so we can leave that for now. 

Might need some changes if we decide on some UI/UX changes further, but looks fine for now.

sahil839 reviewed (commented):
Posted few comments.

sahil839 commented:
For live-updating `avatar_source`, I think you can include the field in `realm_bot` event sent in `notify_avatar_url_change`.

sahil839 commented:
Also, while opening the issue, I thought that we had backend support to remove the avatar from bots and would not be much work, now I realise that we don't have it. So, would be nice to confirm first, whether we actually want to support deleting avatar for bots before moving forward. @alya FYI.

alya commented:
> I think we can support removing avatar for bots like we do for normal users, it is not too much amount of work.

Sure, that makes sense to me.

Ddharmani3 commented:
@sahil839, I have addressed all of your feedback above. However, the approach that you suggested here did not work for updating avatar_source in real-time. In the updated PR, I have kept updating avatar_source in the frontend using bot.avatar_source = result.avatar_source. I believe that this approach is appropriate for testing the new feature's UI/UX aspect. I will create an issue as a follow-up to this PR for someone to work on the backend part of this.
What are your thoughts on this?

sahil839 commented:
> @sahil839, I have addressed all of your feedback above. However, the approach that you suggested [here](https://github.com/zulip/zulip/pull/24697#issuecomment-1482536795) did not work for updating avatar_source in real-time. In the updated PR, I have kept updating avatar_source in the frontend using bot.avatar_source = result.avatar_source. I believe that this approach is appropriate for testing the new feature's UI/UX aspect. I will create an issue as a follow-up to this PR for someone to work on the backend part of this. What are your thoughts on this?

I think passing `avatar_source` to the event should work, but it may require some frontend changes as well.

Ddharmani3 commented:
> I think passing `avatar_source` to the event should work, but it may require some frontend changes as well. Should not be much work. I can try once I have some time.

can you guide me a bit more in this direction? I want to try to solve this issue myself.

sahil839 commented:
The events sent from the server include `avatar_source` field. Now, you need to make sure that on receiving these events, client updates its data correctly from the events. `server_events_dispatch.js` is the starting point for this. Basically you should check whether the data available with client is updated after receiving the events.

Ddharmani3 commented:
@sahil839 I have tried to fix that issue. It is working as expected. can you please check if the approach is correct?

sahil839 reviewed (commented):
Posted few comments. I have not looked at the frontend code completely yet. Will try to do it later today.

Ddharmani3 committed:
realm_user: Add avatar_source field to realm_user.

This is a prep commit to include avatar_source field in realm_user
to enable delete bot avatar in bot edit modal.

Updated format_user_row function in lib/users.py to also
return avatar_source field in realm_user.

Fixes #24427

Ddharmani3 committed:
bots: Add option to remove current bot avatar in edit bot modal.

Added new feature for completely removing user uploaded avatar
by adding a delete button which shows on hovering on current
avatar only if its not a gravtar.

Added a onclick event listener for the delete button in avatar.js
which change the current avatar image to bot gravatar and change
value of a `#current_bot_avatar_source` which is an input of
type hidden to "G".

The value of a `#current_bot_avatar_source` is passed to backend
only when dialog_submit_button in edit bot form is clicked.

Fixes #24427

Ddharmani3 commented:
@sahil839 updated the PR to address feedback. This PR is ready for review now.

sahil839 commented on web/templates/settings/edit_bot_form.hbs:
```diff
@@ -32,9 +32,22 @@
+                <img class="image-block" id="current_bot_avatar_image"  src="{{ bot_avatar_url }}"/>
+                <input type="hidden" id="current_bot_avatar_source"  value="{{ bot_avatar_source }}"/>
+                <div class="image_file_input_error text-error"></div>
+            </div>
```
Can we use `image_upload_widget.hbs` here to avoid duplication?

sahil839 commented on web/src/avatar.js:
```diff
@@ -39,6 +40,23 @@ export function build_bot_edit_widget($target) {
+        $("#edit-bot-avatar-upload-widget .image-hover-background").hide();
+        $("#edit-bot-avatar-upload-widget .image-delete-button").hide();
+        $("#current_bot_avatar_source").val("G");
+        $("#current_bot_avatar_source").trigger("input");
```
I guess this is to enable the save button if it is disabled. But maybe we can do something better, probably just enable the save button directly.

sahil839 reviewed (commented):
Thanks for updating the PR. Posted some comments.
